pkg/operator/resource/resourceapply/volumesnapshotclass.go

```diff
@@ -0,0 +1,91 @@
+package resourceapply
+
+import (
+	"context"
+
+	"k8s.io/klog/v2"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+const (
+	VolumeSnapshotClassGroup    = "snapshot.storage.k8s.io"
+	VolumeSnapshotClassVersion  = "v1"
+	VolumeSnapshotClassResource = "volumesnapshotclasses"
+)
+
+var volumeSnapshotClassResourceGVR schema.GroupVersionResource = schema.GroupVersionResource{
+	Group:    VolumeSnapshotClassGroup,
+	Version:  VolumeSnapshotClassVersion,
+	Resource: VolumeSnapshotClassResource,
+}
+
+func ensureGenericVolumeSnapshotClass(required, existing *unstructured.Unstructured) (*unstructured.Unstructured, bool, error) {
+	requiredParameters, _, err := unstructured.NestedMap(required.UnstructuredContent(), "parameters")
+	if err != nil {
+		return nil, false, err
+	}
+	existingParameters, _, err := unstructured.NestedMap(existing.UnstructuredContent(), "parameters")
+	if err != nil {
+		return nil, false, err
+	}
+
+	if equality.Semantic.DeepEqual(existingParameters, requiredParameters) {
+		return existing, false, nil
+	}
+
+	existingCopy := existing.DeepCopy()
+	if err := unstructured.SetNestedMap(existingCopy.UnstructuredContent(), requiredParameters, "parameters"); err != nil {
+		return nil, true, err
+	}
+
+	return existingCopy, true, nil
+}
+
+// ApplyVolumeSnapshotClass applies Volume Snapshot Class.
+func ApplyVolumeSnapshotClass(client dynamic.Interface, recorder events.Recorder, required *unstructured.Unstructured) (*unstructured.Unstructured, bool, error) {
+	existing, err := client.Resource(volumeSnapshotClassResourceGVR).Get(context.TODO(), required.GetName(), metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		newObj, createErr := client.Resource(volumeSnapshotClassResourceGVR).Create(context.TODO(), required, metav1.CreateOptions{})
+		if createErr != nil {
+			recorder.Warningf("VolumeSnapshotClassCreateFailed", "Failed to create VolumeSnapshotClass.snapshot.storage.k8s.io/v1: %v", createErr)
+			return nil, true, createErr
+		}
+		recorder.Eventf("VolumeSnapshotClassCreated", "Created VolumeSnapshotClass.snapshot.storage.k8s.io/v1 because it was missing")
+		return newObj, true, nil
+	}
+	if err != nil {
+		return nil, false, err
+	}
+
+	existingCopy := existing.DeepCopy()
```

Comment:
Why copy the object here? It gets copied again in `ensureGenericVolumeSnapshotClass`